pkg/pillar/zedcloud/deferred.go

```diff
@@ -24,132 +27,161 @@ import (
+	ticker                 flextimer.FlexTickerHandle
+	priorityCheckFunctions []TypePriorityCheckFunction
+	lock                   *sync.Mutex
+	sentHandler            *SentHandlerFunction
+	zedcloudCtx            *ZedCloudContext
 }
 
+//TypePriorityCheckFunction returns true in case of find type with high priority
+type TypePriorityCheckFunction func(itemType interface{}) bool
+
+//SentHandlerFunction allow doing something with data if it was handled
+//result indicates sending result
+type SentHandlerFunction func(itemType interface{}, data *bytes.Buffer, result types.SenderResult)
+
 // GetDeferredChan creates and returns a channel to the caller
 // We always keep a flextimer running so that we can return
 // the associated channel. We adjust the times when we start and stop
 // the timer.
-func GetDeferredChan(zedcloudCtx *ZedCloudContext) <-chan time.Time {
+// sentHandler is callback which will be run on successful sent
+// priorityCheckFunctions may be added to send item with matched itemType firstly
+// default function at the end of priorityCheckFunctions added to serve non-priority items
+func GetDeferredChan(zedcloudCtx *ZedCloudContext, sentHandler *SentHandlerFunction, priorityCheckFunctions ...TypePriorityCheckFunction) <-chan time.Time {
+	//append with return first
+	priorityCheckFunctions = append(priorityCheckFunctions, func(obj interface{}) bool {
+		return true
+	})
 	zedcloudCtx.deferredCtx = DeferredContext{
-		deferredItems: make(map[string]deferredItemList),
-		ticker:        flextimer.NewRangeTicker(longTime1, longTime2),
+		lock:                   &sync.Mutex{},
+		ticker:                 flextimer.NewRangeTicker(longTime1, longTime2),
+		sentHandler:            sentHandler,
+		priorityCheckFunctions: priorityCheckFunctions,
+		zedcloudCtx:            zedcloudCtx,
 	}
 	return zedcloudCtx.deferredCtx.ticker.C
 }
 
-// Try to send all deferred items. Give up if any one fails
+// HandleDeferred try to send all deferred items (or only one if sendOne set). Give up if any one fails
 // Stop timer if map becomes empty
 // Returns true when there are no more deferred items
-func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration) bool {
+func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.Duration, sendOne bool) bool {
 
-	return zedcloudCtx.deferredCtx.handleDeferred(zedcloudCtx.log, event, spacing)
+	return zedcloudCtx.deferredCtx.handleDeferred(zedcloudCtx.log, event, spacing, sendOne)
 }
 
 func (ctx *DeferredContext) handleDeferred(log *base.LogObject, event time.Time,
-	spacing time.Duration) bool {
+	spacing time.Duration, sendOne bool) bool {
+	ctx.lock.Lock()
+	defer ctx.lock.Unlock()
+
+	if len(ctx.deferredItems) == 0 {
+		return true
+	}
 
-	log.Functionf("HandleDeferred(%v, %v) map %d\n",
+	log.Functionf("HandleDeferred(%v, %v) items %d",
 		event, spacing, len(ctx.deferredItems))
+
 	iteration := 0 // Do some load spreading
```

Comment:
I think this either needs to be an argument or part of the context so that in the normal case when handleDeferred sends the only message we still increment the iteration for each send, as opposed to always using zero.

Reply:
Done